### PR TITLE
Jack

### PR DESCRIPTION
**Pull request by jackraymond**
Banishing raster from lexicon (renaming of modules and functions)
Some minor feature enhancements with respect to tile parameter, with tests implemented.
TO DO: Create a feasibility.py module?
TO DO: Add S as argument to visualize_embeddings?
TO DO: Add more unittests?

### CODE REVIEW

File: minorminer/utils/embeddings.py [hunk 1]
```diff
@@ -131,18 +136,32 @@ def visualize_embeddings(
...
+    Args:
+        G (nx.Graph): A networkx graph
+        prng (np.random.Generator, optional): When provided, is used to shuffle
+            the order of nodes and edges in the source and target graph. This
+            can allow sampling from otherwise deterministic routines.
+    Returns:
+        nx.Graph: The same graph with modified node and edge ordering.
+    """
     if prng is None:
         prng = np.random.default_rng()
```
Comment (kevinchern) [inline minorminer/utils/embeddings.py]: numpy handles `None` automatically
```suggestion
    prng = np.random.default_rng(prng)
```

File: minorminer/utils/embeddings.py [hunk 2]
```diff
@@ -131,18 +136,32 @@ def visualize_embeddings(
...
+    edges = list(G.edges())
     prng.shuffle(edges)
-    _T = nx.Graph()
-    _T.add_nodes_from(nodes)
-    _T.add_edges_from(edges)
-    return _T
+    _G = nx.Graph()
+    _G.add_nodes_from(nodes)
+    _G.add_edges_from(edges)
+    return _G
```
Comment (kevinchern) [inline minorminer/utils/embeddings.py]:
```suggestion
    H = nx.Graph()
    H.add_nodes_from(nodes)
    H.add_edges_from(edges)
    return H
```
or even `G`

File: minorminer/utils/embeddings.py [hunk 3]
```diff
@@ -307,20 +328,19 @@ def embedding_feasibility_filter(
             return True
 
 
-def raster_breadth_subgraph_upper_bound(T: nx.Graph = None) -> int:
-    """Determines a raster breadth upper bound for subgraph embedding.
+def graph_rows_upper_bound(T: nx.Graph = None) -> int:
```
Comment (kevinchern) [inline minorminer/utils/embeddings.py]: what's a "graph rows"?

File: minorminer/utils/embeddings.py [hunk 4]
```diff
@@ -486,35 +515,53 @@ def raster_embedding_search(
...
     else:
         raise ValueError(
             "source graphs must a graph constructed by "
             "dwave_networkx as chimera, pegasus or zephyr type"
         )
-
+    if tile == S:
+        tiling = True
+    else:
+        tiling = False
```
Comment (kevinchern) [inline minorminer/utils/embeddings.py]:
```suggestion
    tiling = tile == S
```

File: tests/utils/test_embeddings.py [hunk 5]
```diff
@@ -241,64 +252,65 @@ def testSubgraphEmbeddingFeasibilityFilter(self):
         )
         # Check tetrahedron cannot be embedded on a graph with a triangle + 0,1,2 connectivity nodes.
 
-    def testRasterBreadthSubgraphUpperBound(self):
+    def testGraphRowsSubgraphUpperBound(self):
```
Comment (kevinchern) [inline tests/utils/test_embeddings.py]: snake case for this and other test functions

File: tests/utils/test_embeddings.py [hunk 6]
```diff
@@ -314,6 +326,41 @@ def testRasterEmbeddingSearchBasic(self):
...
+        T = dnx.chimera_graph(min_graph_rows + 1)
+        tile = dnx.chimera_graph(min_graph_rows, node_list=list(range(1, 8)))
+        embs = find_sublattice_embeddings(
+            S, T, graph_rows=min_graph_rows, max_num_emb=float("Inf"), tile=tile
+        )
+        self.assertEqual(len(embs), 4)
+        nodes_used = {v for emb in embs for v in emb.values()}
+        self.assertEqual(len(nodes_used), S.number_of_nodes() * len(embs))
+        self.assertTrue(
+            all(n % 8 > 0 for n in nodes_used), "Every 8th node excluded by tile"
```
Comment (kevinchern) [inline tests/utils/test_embeddings.py]: is the error message correct? (and below)

File: tests/utils/test_embeddings.py [hunk 7]
```diff
@@ -314,6 +326,41 @@ def testRasterEmbeddingSearchBasic(self):
...
+        )
+        tile5 = dnx.chimera_graph(min_graph_rows, node_list=list(range(3, 8)))
+        embs = find_sublattice_embeddings(
+            S, T, graph_rows=min_graph_rows, max_num_emb=float("Inf"), tile=tile5
+        )
+        self.assertEqual(len(embs), 0, "Tile is too small")
+
+        S = tile
```
Comment (kevinchern) [inline tests/utils/test_embeddings.py]: break into multiple tests

File: tests/utils/test_embeddings.py [hunk 8]
```diff
@@ -314,6 +326,41 @@ def testRasterEmbeddingSearchBasic(self):
                 len(set(value_list)), len(value_list), "embeddings overlap"
             )
 
+    def testFindSublatticeEmbeddingsTile(self):
```
Comment (kevinchern) [inline tests/utils/test_embeddings.py]: more specific name

File: tests/utils/test_embeddings.py [hunk 9]
```diff
@@ -241,64 +252,65 @@ def testSubgraphEmbeddingFeasibilityFilter(self):
...
         self.assertEqual(
-            m // 2, raster_breadth_subgraph_lower_bound(S=S, topology="zephyr")
+            m // 2, graph_rows_lower_bound(S=S, topology="zephyr", one_to_one=True)
         )
         T = dnx.zephyr_graph(m)
-        self.assertEqual(m // 2, raster_breadth_subgraph_lower_bound(S=S, T=T))
+        self.assertEqual(m // 2, graph_rows_lower_bound(S=S, T=T, one_to_one=True))
 
-    def testRasterEmbeddingSearchBasic(self):
+    def testFindSublatticeEmbeddingsBasic(self):
```
Comment (kevinchern) [inline tests/utils/test_embeddings.py]: should these tests account for edge cases? e.g., invalid input

File: minorminer/utils/embeddings.py [hunk 10]
```diff
@@ -529,17 +576,22 @@ def raster_embedding_search(
...
-            Tr,
-            max_num_emb=max_num_emb,
-            timeout=timeout,
-            skip_filter=skip_filter,
-            prng=prng,
-            embedder=embedder,
-            embedder_kwargs=embedder_kwargs,
-        )
+        if tiling:
+            if Tr.number_of_edges() == S.number_of_edges():
```
Comment (kevinchern) [inline minorminer/utils/embeddings.py]: add comment to explain why this special case exists.
probably good idea to briefly describe other cases too.

**Review by kevinchern — CHANGES_REQUESTED** [review]

File: minorminer/utils/embeddings.py (hunk 1 above)
Comment (jackraymond) [reply minorminer/utils/embeddings.py]: We should change the name of the input argument to seed. Allow types: Union[int, np.random.RandomState, np.random.Generator], and document that the seed can be an integer used to seed a numpy random generator,  or a numpy random number generator."

File: minorminer/utils/embeddings.py (hunk 3 above)
Comment (jackraymond) [reply minorminer/utils/embeddings.py]: The parameter 'm' of Chimera, Pegasus and Zephyr graphs is equal to G.graph['rows'] and is described in documentation as number of rows. The raster breadth is the value of m describing the subgraph. I think we can improve the documentation, but I'm struggling for a better choice than graph rows.

File: tests/utils/test_embeddings.py (hunk 6 above)
Comment (jackraymond) [reply tests/utils/test_embeddings.py]: On each tile the 8th node is excluded. So the nodes_used must excluded 8, 16, 24, 32. If the comment is not clear lets delete it. Otherwise can you make a recommendation.

File: tests/utils/test_embeddings.py (hunk 7 above)
Comment (jackraymond) [reply tests/utils/test_embeddings.py:351]: TestTile1 and TestTile2? Some of the same objects are being used throughout, and its only a test of one input argument, therefore I was thinking it was appropriate to keep all the tile variation tests together.

Comment (jackraymond) [reply tests/utils/test_embeddings.py]: Wouldn't it make sense to keep all the tests associated to the input argument 'tile' together, especially since the tests share input object and context.

File: tests/utils/test_embeddings.py (hunk 5 above)
Comment (jackraymond) [reply tests/utils/test_embeddings.py]: Yep, its Classes that should have Camel case, good catch

File: tests/utils/test_embeddings.py (hunk 8 above)
Comment (jackraymond) [reply tests/utils/test_embeddings.py]: testTileInputArgument ?

File: tests/utils/test_embeddings.py (hunk 9 above)
Comment (jackraymond) [reply tests/utils/test_embeddings.py]: We need to add self.assertRaises() tests, but beyond that I'm not sure we need to check every single wrong type throws an error given that we are type hinting

File: minorminer/utils/embeddings.py (hunk 10 above)
Comment (jackraymond) [reply minorminer/utils/embeddings.py]: Comment in the docstring isn't good enough?

**Comment by jackraymond** [comment]
A couple of changes:

Per @Kevin's suggestion above. We should change the prng argument to seed, and then create the prng from the seed. This is tidier and standard for ocean-sdk.

Rather than breaking this module into two parts as had been brainstormed, perhaps leave as is. The filter functions, min_graph_rows function, are not worldbeaters requiring their own module. We can raise the possibility on pull request though. This gets us to the pull request more quickly 

We have a timeout parameter that is defaulted somewhat arbitrarily. Since we now have an embedder_kwargs I think it would be better to omit this, since it is actually a property of the subsolver. I also think we should rely on the subsolver default, rather than introducing our own. We can warn the user in the docstring that unless timeouts/ max_num_embs are set the code can be slow.
 
We haven't tested  'trigger_restarts=True' for subgraph, or other possible non-defaulted arguments. I think it would be better to remove this and rely on subgraph.find_embeddings defaults:  https://github.com/ciaranm/glasgow-subgraph-solver/blob/master/src/glasgow_subgraph_solver.cc
If time permits, we can test some of these parameters, but I think what is really required is for someone to revisit the subgraph documentation first (Kevin has raised an issue in ocean-sdk)

We need to add a Raises to all our docstrings that raise errors. We also need to add self.assertRaises tests.
https://google.github.io/styleguide/pyguide.html?showone=Imports_formatting#383-functions-and-methods

File: minorminer/utils/embeddings.py (hunk 1 above)
Comment (AndyZzzZzzZzz) [reply minorminer/utils/embeddings.py]: I used to put more detailed type hinting but somethings it makes the code too long.

File: minorminer/utils/embeddings.py (hunk 2 above)
Comment (AndyZzzZzzZzz) [reply minorminer/utils/embeddings.py]: I think H was usued in the original visualize subgraph function inside Ocean.

File: minorminer/utils/embeddings.py (hunk 3 above)
Comment (jackraymond) [reply minorminer/utils/embeddings.py]: I will improve the document if you two don't have any specific suggestions.
We could try graph_rows_m (to make a connection with the shape parameter m)?
sublattice_rows ?
sublattice_rows_m ?
sublattice_dimension?
sublattice_scale

Comment (kevinchern) [reply minorminer/utils/embeddings.py]: I like descriptive names like `sublattice_size` or `sublattice_scale` or `sublattice_dim`

File: minorminer/utils/embeddings.py (hunk 10 above)
Comment (kevinchern) [reply minorminer/utils/embeddings.py]: Ooops, missed it in the docstring.

File: tests/utils/test_disjoint_embeddings.py [hunk 11]
```diff
@@ -314,6 +346,90 @@ def testRasterEmbeddingSearchBasic(self):
...
+        nodes_used = {v for emb in embs for v in emb.values()}
+        self.assertEqual(len(nodes_used), S.number_of_nodes() * len(embs))
+
+        invalid_T = nx.complete_graph(5)  # Complete graph is not a valid topology
+        with self.assertRaises(ValueError):
+            find_sublattice_embeddings(
+                S, invalid_T, graph_rows=min_graph_rows, tile=tile
+            )
+
+        # small_T = dnx.chimera_graph(2,2)
```
Comment (AndyZzzZzzZzz) [inline tests/utils/test_disjoint_embeddings.py]: This test is still not working. Goal: create a source graph that is small enough to fit into target graph but too large for tiles so itr triggers corresponding raises.

File: minorminer/utils/disjoint_embeddings.py [hunk 12]
```diff
@@ -329,29 +378,33 @@ def raster_breadth_subgraph_lower_bound(
...
-            None if embedding for any raster breadth is infeasible.
+
+    Raises:
+        ValueError: If `T` and `topology` are inconsistent, or if `T` and `t` are inconsistent.
+        ValueError: If neither `T` nor `topology` is specified.
+        ValueError: If `S` cannot be embedded in `T` or the specified topology.
+
+    Returns:
+        float: Minimum `graph_rows` for embedding to be feasible. Returns
+            None if embedding for any number of graph rows is infeasible.
```
Comment (AndyZzzZzzZzz) [inline minorminer/utils/disjoint_embeddings.py]: Line 413 - 423 contains a duplicate check for topology != T family:
```
  if T is not None:
        if embedding_feasibility_filter(S, T, one_to_one) is False:
            return None
        if topology is None:
            topology = T.graph.get("family")
        elif topology != T.graph.get("family"):
            raise ValueError("Arguments T and topology are inconsistent")
        if t is None:
            t = T.graph["tile"]
        elif topology != T.graph.get("family"):
            raise ValueError("Arguments T and t are inconsistent")

File: tests/utils/test_feasibility.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+from minorminer.utils.feasibility import (
+    embedding_feasibility_filter,
+    lattice_size_upper_bound,
+    lattice_size_lower_bound,
+)
+
+
+class TestEmbeddings(unittest.TestCase):
+    def test_subgraph_embedding_feasibility_filter(self):
```
Comment (jackraymond) [inline tests/utils/test_feasibility.py]:
```suggestion
    def test_embedding_feasibility_filter(self):
```
